llvm/lib/Transforms/IPO/FunctionSpecialization.cpp

```diff
@@ -682,10 +679,9 @@ bool FunctionSpecializer::run() {
         (RequireMinSize && Metrics.NumInsts < MinFunctionSize))
       continue;
 
-    // TODO: For now only consider recursive functions when running multiple
-    // times. This should change if specialization on literal constants gets
-    // enabled.
-    if (!Inserted && !Metrics.isRecursive && !SpecializeLiteralConstant)
+    // When specialization on literal constants is disabled, only consider
+    // recursive functions when running multiple times.
+    if (!SpecializeLiteralConstant && !Inserted && !Metrics.isRecursive)
```

Comment:
Ah yep thank you for clarifying this point, I have updated the comment to be clearer on this point. I reordered the checks because it seemed to make more sense to check SpecializeLiteralConstant=false, given that it is only in this case that we limit multiple iterations to recursive functions.

On this point, I was actually thinking - is it not hypothetically possible that a specialized function could also be found to return a constant _pointer_, e.g. a function pointer?

Reply:
Hmm, that did not occur to me before good point. Would it be too much to ask adding a regression test? :) We already have one for tracking return values, perhaps you could use that as a template?
https://github.com/llvm/llvm-project/blob/main/llvm/test/Transforms/FunctionSpecialization/track-return.ll

Reply:
No problem - I've added https://github.com/llvm/llvm-project/pull/113442/commits/4533583136a11f260b391c0bf9be2aeb72b389fc#diff-43f6ba03f40b792b303a3a0a05b40dccb7fe9898a649e93d71038be1ee6ebb0c

I've intentionally written it as a motivating case for removing this condition altogether, as it is an example of a non-recursive function that could be specialized on a successive iteration when `SpecializeConstantLiteral=false`.

Do you think that this makes? I don't believe it will have a significant impact on compile times - you could perhaps argue that my example is a tad contrived, however...